plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -534,6 +666,68 @@ private static ConnectorPageSource createParquetPageSource(
+     */
+    public static Optional<ReaderColumns> projectColumns(List<IcebergColumnHandle> columns)
+    {
+        requireNonNull(columns, "columns is null");
+
+        // No projection is required if all columns are base columns
+        if (columns.stream().allMatch(IcebergColumnHandle::isBaseColumn)) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<ColumnHandle> projectedColumns = ImmutableList.builder();
+        ImmutableList.Builder<Integer> outputColumnMapping = ImmutableList.builder();
+        Map<Integer, Integer> mappedFieldIds = new HashMap<>();
+        int projectedColumnCount = 0;
+
+        for (IcebergColumnHandle column : columns) {
+            int baseColumnId = column.getBaseColumnIdentity().getId();
+            Integer mapped = mappedFieldIds.get(baseColumnId);
+
+            if (mapped == null) {
+                projectedColumns.add(column.getBaseColumn());
+                mappedFieldIds.put(baseColumnId, projectedColumnCount);
+                outputColumnMapping.add(projectedColumnCount);
+                projectedColumnCount++;
+            }
+            else {
+                outputColumnMapping.add(mapped);
+            }
+        }
+
+        return Optional.of(new ReaderColumns(projectedColumns.build(), outputColumnMapping.build()));
+    }
+
+    /**
+     * Maps the Iceberg field id to the Parquet schema type. For nested columns this includes the necessary GroupTypes for the reader to reference the nested column.
+     * Nested types are separated so they can be independently referenced for projection pushdown by their individual iceberg field ids.
+     */
+    private static Map<Integer, org.apache.parquet.schema.Type> mapIdsToParquetType(List<org.apache.parquet.schema.Type> fields)
+    {
+        ImmutableMap.Builder<Integer, org.apache.parquet.schema.Type> columnMap = ImmutableMap.builder();
+        fields.stream()
+                .filter(field -> field.getId() != null)
+                .forEach(field -> {
+                    // Add one entry for the top level type
+                    columnMap.put(field.getId().intValue(), field);
+                    if (!field.isPrimitive()) {
+                        // Add an entry for each of the children, while preserving the nested structure
+                        GroupType groupType = field.asGroupType();
+                        columnMap.putAll(mapIdsToParquetType(groupType.getFields())
+                                .entrySet()
+                                .stream()
+                                .collect(toMap(
+                                        Map.Entry::getKey,
+                                        // Preserve the parent type details so the schema can be constructed from the root of the type tree
+                                        entry -> new GroupType(field.getRepetition(), field.getName(), ImmutableList.of(entry.getValue())))));
```

Comment:
I don't understand what it does, but the final result feels hacky: in the returned map under some ID key K we don't have the parquet Type with Id K.

It could be that https://github.com/trinodb/trino/pull/9104 solves this more elegantly, by changing `ParquetColumnIOConverter`.

Reply:
This method is actually unnecessary now that we're projecting based off of the base column all the time. I'll roll this change back.